glean-core/rlb/src/dispatcher/mod.rs

```diff
@@ -247,3 +273,288 @@ impl Dispatcher {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::sync::atomic::{AtomicBool, AtomicU8, Ordering};
+    use std::sync::{Arc, Mutex};
+    use std::{thread, time::Duration};
+
+    fn enable_test_logging() {
+        // When testing we want all logs to go to stdout/stderr by default,
+        // without requiring each individual test to activate it.
+        let _ = env_logger::builder().is_test(true).try_init();
+    }
+
+    #[test]
+    fn tasks_run_off_the_main_thread() {
+        enable_test_logging();
+
+        let main_thread_id = thread::current().id();
+        let thread_canary = Arc::new(AtomicBool::new(false));
+
+        let mut dispatcher = Dispatcher::new(100);
+
+        // Force the Dispatcher out of the pre-init queue mode.
+        dispatcher
+            .flush_init()
+            .expect("Failed to get out of preinit queue mode");
+
+        let canary_clone = thread_canary.clone();
+        dispatcher
+            .guard()
+            .launch(move || {
+                assert!(thread::current().id() != main_thread_id);
+                // Use the canary bool to make sure this is getting called before
+                // the test completes.
+                assert_eq!(false, canary_clone.load(Ordering::SeqCst));
+                canary_clone.store(true, Ordering::SeqCst);
+            })
+            .expect("Failed to dispatch the test task");
+
+        dispatcher.block_on_queue();
+        assert_eq!(true, thread_canary.load(Ordering::SeqCst));
+        assert_eq!(main_thread_id, thread::current().id());
+    }
+
+    #[test]
+    fn launch_correctly_adds_tasks_to_preinit_queue() {
+        enable_test_logging();
+
+        let thread_canary = Arc::new(AtomicU8::new(0));
+
+        let mut dispatcher = Dispatcher::new(100);
+
+        // Add 3 tasks to queue each one increasing thread_canary by 1 to
+        // signal that the tasks ran.
+        for _ in 0..3 {
+            let canary_clone = thread_canary.clone();
+            dispatcher
+                .guard()
+                .launch(move || {
+                    canary_clone.fetch_add(1, Ordering::SeqCst);
+                })
+                .expect("Failed to dispatch the test task");
+        }
+
+        // Ensure that no task ran.
+        assert_eq!(0, thread_canary.load(Ordering::SeqCst));
+
+        // Flush the queue and wait for the tasks to complete.
+        dispatcher
+            .flush_init()
+            .expect("Failed to get out of preinit queue mode");
+        dispatcher.block_on_queue();
```

Comment:
This shouldn't be necessary because flush_init blocks on flushing the preinit operations, right?